### PR TITLE
VPC template.

### PR DESCRIPTION
**Pull request by JanDoerrenhaus**

### CODE REVIEW

File: src/main/template/vpc.yaml
Comment (schroenser) [inline src/main/template/vpc.yaml]: Technically, we _could_ perform this arithmetic operation using a custom resource (see [Stackoverflow](https://stackoverflow.com/a/41422057/2060692)). However, this was deemed "bonus points" for this issue, so no action needed.

File: src/main/template/vpc.yaml
Comment (schroenser) [inline src/main/template/vpc.yaml:30]: I wonder if we need or even should support adding a key from the AWS Key manager, which is obviously an account key, when we usually discourage the use of non-personal keys and register a bunch of personal keys via `authorizedKeys` anyways. The `KeyName` property on an EC2 instance is not required.

File: src/main/template/vpc.yaml
Comment (schroenser) [inline src/main/template/vpc.yaml:228]: [The documentation for the `Domain` property](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-eip.html#aws-properties-ec2-eip-properties) states that
> If you define an Elastic IP address and associate it with a VPC that is defined in the same template, you must declare a dependency on the VPC-gateway attachment by using the DependsOn Attribute on this resource.

I wonder why this worked for us in the past. Am I missing something?

File: src/main/template/vpc.yaml
Comment (schroenser) [inline src/main/template/vpc.yaml]: We also need to output the `GroupId` attribute of the `bastionSecurityGroup` so other security groups can later use it in their `SecurityGroupIngress` object's `SourceSecurityGroupId` properties. See [the examples section of the `SecurityGroupIngress` object](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-security-group-ingress.html#aws-properties-ec2-security-group-ingress--examples) for details.

File: src/main/template/vpc.yaml
Comment (schroenser) [inline src/main/template/vpc.yaml]: [The documentation on the `NetworkInterfaces` property](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-instance.html#aws-properties-ec2-instance-properties) states that
> If this resource has a public IP address and is also in a VPC that is defined in the same template, you must use the DependsOn Attribute to declare a dependency on the VPC-gateway attachment.

Same question: why does this currently work? O_o

**Review by schroenser — CHANGES_REQUESTED** [review]

File: src/main/template/vpc.yaml
Comment (JanDoerrenhaus) [reply src/main/template/vpc.yaml:228]: Because this is a race condition. If this elastic IP were to be created before the VPC gateway attachment, it would fail. In the current order of things, this does not happen. But by luck, not by enforcement.

File: src/main/template/vpc.yaml
Comment (JanDoerrenhaus) [reply src/main/template/vpc.yaml]: See above.

File: src/main/template/vpc.yaml
Comment (JanDoerrenhaus) [reply src/main/template/vpc.yaml:30]: Yes, we should. Placing this key on the instance will always work. Your custom setup which adds more keys might not. If it didn't, you can still diagnose the machine by using the account key. Otherwise, you have no way of telling what went wrong.

File: src/main/template/vpc.yaml
Comment (schroenser) [inline src/main/template/vpc.yaml]: Replace `it.bastionSecurityGroup` with `it.`.

File: src/main/template/vpc.yaml
Comment (schroenser) [inline src/main/template/vpc.yaml]: LOL. I missed this during my first review round. Good catch.

**Review by schroenser — CHANGES_REQUESTED** [review]

**Comment by schroenser** [comment]
Added supreme judge now to reduce overall review time.

File: src/main/template/vpc.yaml
Comment (bannmann) [inline src/main/template/vpc.yaml:11]: a *larger* subnet would actually have a subnet mask of e.g. 8 bits. in this context, "larger than 16 bits" is ambiguous as it may also be read as e.g. "24 bits" (more numbers, but a *smaller* subnet). sure, the reader should be familiar with subnetting, but we should still try to improve the wording here.

Comment (bannmann) [reply src/main/template/vpc.yaml:11]: reading a bit more about the topic, I also guess that readers might confuse the bit value here with the third argument of `Fn:Cidr`, where a larger value actually means a larger subnet:

> `cidrBits`
>    The number of subnet bits for the CIDR. For example, specifying a value "8" for this parameter will create a CIDR with a mask of "/24".
>    **Note:** Subnet bits is the inverse of subnet mask. To calculate the required host bits for a given subnet bits, subtract the subnet bits from 32 for IPv4 or 128 for IPv6.

File: src/main/template/vpc.yaml
Comment (bannmann) [inline src/main/template/vpc.yaml]: we should rename this from "partition bits" (a made-up term?) to "cidr bits" (the name of the `Fn:Cidr` parameter we use this value for).

also, do we really need to allow different values for the individual quarters? this feels YAGNI to me.

File: src/main/template/vpc.yaml
Comment (bannmann) [inline src/main/template/vpc.yaml:40]: add "on all machines".

Comment (bannmann) [reply src/main/template/vpc.yaml:40]: Now I realize this is only about the bastion, of course. I suggest renaming the parameter to `bastionAuthorizedKeys` and keeping the description as is.

File: src/main/template/vpc.yaml
Comment (bannmann) [inline src/main/template/vpc.yaml:394]: any reason for `2`? I would have expected `1`.

File: src/main/template/vpc.yaml
Comment (bannmann) [inline src/main/template/vpc.yaml:414]: What's the rationale for writing `Name` tags of all resources with spaces (`MyGreatStack Private Route Table C`), but names of outputs with dashes (`MyGreatStack-public-subnets`)? Does it make sense to unify them one way or the other?

**Review by bannmann — CHANGES_REQUESTED** [review]

File: src/main/template/vpc.yaml
Comment (schroenser) [reply src/main/template/vpc.yaml:394]: If the `MaxSize` is only `1`, the autoscaling group will wait for the instance to be completely terminated before spinning up a new one. Having it at `2` allows for a new instance to be started as soon as the first one is deemed dead, regardless of the termination process.

File: src/main/template/vpc.yaml
Comment (schroenser) [reply src/main/template/vpc.yaml]: Note that those aren't for individual quarters but for the first and second quartering calcuation, where the second one results in eighths of the whole. This is why we need different values for both.
Your misinterpretation is a good example for the fact that the "user" of this template might not know what is expected of him. In fact, he can't choose freely. `cidrBlock`, `firstQuarteringPartitionBits` and `secondQuarteringPartitionBits` _have_ adhere to the correct mathematics, otherwise weird stuff will happen.
Perhaps, the best way is indeed to omit those parameters altogether and implement the arithmetic operations via a Lambda and a custom resource, like [my comment above](https://github.com/mizool/cloudformation/pull/1#discussion_r346392617) hints toward.
I simply was not sure if we need to add the complexity for that _now_ or if we can refactor that later.

File: src/main/template/vpc.yaml
Comment (bannmann) [reply src/main/template/vpc.yaml]: With [schroenser's explanation above](https://github.com/mizool/cloudformation/pull/1/files#r347250720) in mind, I suggest replacing the description with this:

> The cidr bits for the four subnets on the first level ("private subnet A", "private subnet B", "private subnet C", "subnet containing public subnets A to C").

We shouldn't document the rationale about missing support of arithmetics in the description (audience: person creating a stack using the template) - if we want, we can do that in a YAML comment (audience: reader/maintainer). I tend towards not documenting it.

File: src/main/template/vpc.yaml
Comment (bannmann) [inline src/main/template/vpc.yaml]: With [schroenser's explanation above](https://github.com/mizool/cloudformation/pull/1/files#r347250720) in mind, I suggest replacing the description with this:

> The cidr bits for the four subnets on the second level ("public subnet A", "public subnet B", "public subnet C", "unused").

File: src/main/template/vpc.yaml
Comment (bannmann) [reply src/main/template/vpc.yaml]: Implementing arithmetics via the lambda approach is too much complexity for now as the use case "but I need to tweak my private IP ranges arbitrarily" is YAGNI.

I think improving the clarity via naming and descriptions is sufficient; see my other comments.

**Review by bannmann — CHANGES_REQUESTED** [review]

File: src/main/template/vpc.yaml
Comment (JanDoerrenhaus) [reply src/main/template/vpc.yaml:414]: The name tags are meant as speaking names to an AWS console user. The export names are identifiers to be used by other stacks.

File: src/main/template/vpc.yaml
Comment (bannmann) [inline src/main/template/vpc.yaml]: this and other refs still need to be updated. please verify things work by creating a stack.

**Review by bannmann — CHANGES_REQUESTED** [review]

File: src/main/template/vpc.yaml
Comment (JanDoerrenhaus) [reply src/main/template/vpc.yaml]: Done. Dammit. Always do these things with search and replace...

**Review by bannmann — APPROVED** [review]

**Comment by bannmann** [comment]
@schroenser, please approve and merge when you're happy with this.

**Review by schroenser — APPROVED** [review]

**Pull request merged**
